packages/devtools_app/lib/src/charts/chart.dart

```diff
@@ -0,0 +1,588 @@
+        drawXTick(canvas, timestamp, xCoord, axis, displayTime: true);
+      }
+    }
+
+    // Rescale Y-axis to max visible Y range.
+    chartController.resetYMaxValue(visibleYMax);
+
+    // Draw Y-axis ticks and labels.
+    if (chartController.displayYLabels) {
+      drawYTicks(canvas, chartController, axis);
+    }
+
+    drawTitle(canvas, size, chartController.title);
+  }
+
+  void clipChart(Canvas canvas, {ClipOp op = ClipOp.intersect}) {
+    final leftSideSide = chartController.xCanvasChart;
+    final topChartSide = chartController.yCanvasChart;
+    final r = Rect.fromLTRB(
+      leftSideSide,
+      topChartSide,
+      chartController.canvasChartWidth +
+          leftSideSide -
+          chartController.xPaddingRight,
+      topChartSide + chartController.canvasChartHeight,
+    );
+
+    canvas.clipRect(r, clipOp: op);
+  }
+
+  // TODO(terry): Use drawText?
+  void drawTitle(Canvas canvas, Size size, String title) {
+    final tp = createText(title, 1.5);
+    tp.paint(canvas, Offset(size.width / 2 - tp.width / 2, 0));
+  }
+
+  void drawAxes(Canvas canvas, Size size, Paint axis) {
+    // Top line of chart.
+    canvas.drawLine(
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart,
+      ),
+      Offset(
+        chartController.canvasChartWidth +
+            chartController.xCanvasChart -
+            chartController.xPaddingRight,
+        chartController.yCanvasChart,
+      ),
+      axis,
+    );
+
+    // Left-side of chart
+    canvas.drawLine(
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart,
+      ),
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart + chartController.canvasChartHeight,
+      ),
+      axis,
+    );
+
+    // Bottom line of chart.
+    canvas.drawLine(
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart + chartController.canvasChartHeight,
+      ),
+      Offset(
+        chartController.canvasChartWidth +
+            chartController.xCanvasChart -
+            chartController.xPaddingRight,
+        chartController.yCanvasChart + chartController.canvasChartHeight,
+      ),
+      axis,
+    );
+  }
+
+  /// Separated out from drawAxis because we don't know range until plotted.
+  // TODO(terry): Consider combining into drawAxis function.
+  void drawYTicks(Canvas canvas, ChartController chartController, Paint axis) {
+    final xCoord = chartController.xCanvasChart;
+
+    final yScale = chartController.yScale;
+
+    for (var labelIndex = yScale.labelTicks; labelIndex >= 0; labelIndex--) {
+      final unit = pow(10, yScale?.labelUnitExponent).toInt();
+      final y = labelIndex * unit;
+      final yCoord = chartController.yPositonToYCanvasCoord(y);
+
+      final labelName = constructLabel(
+        labelIndex.toInt(),
+        yScale.labelUnitExponent.toInt(),
+      );
+
+      drawText('$labelName', canvas, 5, yCoord);
```

Comment:
@pq do we have a lint for unneccssary string comprehensions? Would seem easy to have a lint that realized that
`'$labelName'` could just be `labelName` particularly once we have null safety so there isn't a risk the code is needed to ensure we have the string `'null`'

Reply:
https://dart-lang.github.io/linter/lints/unnecessary_string_interpolations.html 👍

Reply:
Fixed.